src/shared/GitLab/GitLabAuthentication.cs

```diff
@@ -0,0 +1,160 @@
+using GitCredentialManager;
+using GitCredentialManager.Authentication;
+using GitCredentialManager.Authentication.OAuth;
+
+namespace GitLab
+{
+    public interface IGitLabAuthentication : IDisposable
+    {
+        AuthenticationPromptResult GetAuthentication(Uri targetUri, string userName, AuthenticationModes modes);
+
+        Task<OAuth2TokenResult> GetOAuthTokenViaBrowserAsync(Uri targetUri, IEnumerable<string> scopes);
+
+        Task<OAuth2TokenResult> GetOAuthTokenViaRefresh(Uri targetUri, string refreshToken);
+    }
+
+    public class AuthenticationPromptResult
+    {
+        public AuthenticationPromptResult(AuthenticationModes mode)
+        {
+            AuthenticationMode = mode;
+        }
+
+        public AuthenticationPromptResult(AuthenticationModes mode, ICredential credential)
+            : this(mode)
+        {
+            Credential = credential;
+        }
+
+        public AuthenticationModes AuthenticationMode { get; }
+
+        public ICredential Credential { get; set; }
+    }
+
+    [Flags]
+    public enum AuthenticationModes
+    {
+        None = 0,
+        Basic = 1,
+        Browser = 1 << 1,
+        Pat = 1 << 2,
+
+        All = Basic | Browser | Pat
+    }
+
+    public class GitLabAuthentication : AuthenticationBase, IGitLabAuthentication
+    {
+        public GitLabAuthentication(ICommandContext context)
+            : base(context) { }
+
+        public AuthenticationPromptResult GetAuthentication(Uri targetUri, string userName, AuthenticationModes modes)
+        {
+            // If we don't have a desktop session/GUI then we cannot offer browser
+            if (!Context.SessionManager.IsDesktopSession)
+            {
+                modes = modes & ~AuthenticationModes.Browser;
+            }
+
+            // We need at least one mode!
+            if (modes == AuthenticationModes.None)
+            {
+                throw new ArgumentException(@$"Must specify at least one {nameof(AuthenticationModes)}", nameof(modes));
+            }
+
+            switch (modes)
+            {
+                case AuthenticationModes.Basic:
+                case AuthenticationModes.Pat:
+                    ThrowIfUserInteractionDisabled();
+                    ThrowIfTerminalPromptsDisabled();
+                    Context.Terminal.WriteLine("Enter GitLab credentials for '{0}'...", targetUri);
+
+                    if (string.IsNullOrWhiteSpace(userName))
+                    {
+                        userName = Context.Terminal.Prompt("Username");
+                    }
+                    else
+                    {
+                        Context.Terminal.WriteLine("Username: {0}", userName);
+                    }
+
+                    string password_or_token = Context.Terminal.PromptSecret(modes == AuthenticationModes.Basic ? "Password" : "Personal access token");
+                    return new AuthenticationPromptResult(modes, new GitCredential(userName, password_or_token));
```

Comment:
There's a bug here.

You'll need to split these cases in to their own block because `modes` is a flags enumeration, meaning it could be (and is by default) `All`. The returned `AuthenticationPromptResult.AuthenticationMode` is not valid (it should be one flag only; the selected one).

At the same time, the `Context.Terminal.PromptSecret(string)` call always prints `"Personal access token"` because the `modes == .Basic` test always returns `false` (it's `.All`).

I know it's more verbose, and with some duplication of the username/prompt handling, but it's much easier to reason about 😁 

```suggestion
                case AuthenticationModes.Basic:
                    ThrowIfUserInteractionDisabled();
                    ThrowIfTerminalPromptsDisabled();
                    Context.Terminal.WriteLine("Enter GitLab credentials for '{0}'...", targetUri);

                    if (string.IsNullOrWhiteSpace(userName))
                    {
                        userName = Context.Terminal.Prompt("Username");
                    }
                    else
                    {
                        Context.Terminal.WriteLine("Username: {0}", userName);
                    }

                    string password = Context.Terminal.PromptSecret("Password");
                    return new AuthenticationPromptResult(AuthenticationModes.Basic, new GitCredential(userName, password));

                case AuthenticationModes.Pat:
                    ThrowIfUserInteractionDisabled();
                    ThrowIfTerminalPromptsDisabled();
                    Context.Terminal.WriteLine("Enter GitLab credentials for '{0}'...", targetUri);

                    if (string.IsNullOrWhiteSpace(userName))
                    {
                        userName = Context.Terminal.Prompt("Username");
                    }
                    else
                    {
                        Context.Terminal.WriteLine("Username: {0}", userName);
                    }

                    string token = Context.Terminal.PromptSecret("Personal access token");
                    return new AuthenticationPromptResult(AuthenticationModes.Pat, new GitCredential(userName, token));
```

Reply:
Because of the goto! Thanks, I get it!

Reply:
Added a test. goto is not easy to reason about!